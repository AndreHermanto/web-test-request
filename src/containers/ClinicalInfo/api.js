/**
 * Upload an attachment
 * @param {Object} data Test request data to be submitted.
 */
export function upload(file) {
  var formData = new FormData();
  formData.append('file', file);
  return fetch(`${process.env.REACT_APP_BASE_URL}/attachments`, {
    method: 'POST',
    body: formData
  }).then((response) => response.json() );
}
