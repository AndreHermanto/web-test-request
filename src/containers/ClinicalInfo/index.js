import React, { Component } from 'react';
import {
  initData,
  setFormData,
  validatedToTrue
} from './reducer'
import { 
  PageHeading,
  FormButton
} from './../../components/SharedStyle';
import TextArea from './../../components/TextArea';
import Toggle from './../../components/Toggle';

/**
 * PatientDetails - UI for input patient details.
 */
class ClinicalInfo extends Component {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
    this.handleBack = this.handleBack.bind(this);
    this.handleConfirm = this.handleConfirm.bind(this);
    this.state = initData(props.route.data);
  }
  
  componentDidMount() {
    if(this.props.route.preventUnvisitedFormAccess) {
      this.props.router.setRouteLeaveHook(this.props.route, this.props.route.preventUnvisitedFormAccess.bind(this));
    }
  }
  
  handleChange(event) {
    this.setState(setFormData(this.state, event.target));
  }
  
  handleBack() {
    this.props.route.onChange(this);
    this.props.router.push('/step2');
  }
  
  handleNext(passValidation) {
    if(!passValidation) return false;
    if(this.props.route.isEdited === true)
    {
      this.props.route.onChange(this);
      this.props.router.push('/summary')
    }
    else {
      this.props.route.onChange(this);
      this.props.router.push('/step4'); 
    }
  }
  
  handleConfirm() {
    return this.setState(validatedToTrue(this.state), () => {    
      var pass = true;
      for (var field in this.state.validation) {
        if(this.state.validation[field].status === 'error') pass = false;
      }
      this.handleNext(pass); 
    });
  }
  
  // This renders the validation result after confirm button is clicked.
  validate() {
    return this.state.validated && this.state.validation;
  }
  
  render() {
    return (
      <div>
        <PageHeading>Step 3: Clinical Information</PageHeading>
        <TextArea
          field="clinicalInfo"
          label="Provide Clinical Information"
          helper="Type or copy and paste clinical notes and the diagnosis here. Comprehensive notes increase the chance of a successful diagnosis."
          onChange={this.handleChange}
          onValidate={this.validate()}
          formState={this.state.form}
          required
        />

        <TextArea
          field="relevantInvestigation"
          label="Provide results from relevant investigations"
          helper="Provide notes from genetic tests, imaging results, etc."
          onChange={this.handleChange}
          onValidate={this.validate()}
          formState={this.state.form}
          optional
        />
            
        <TextArea
          field="familyHistory"
          label="Family history"
          helper="Please describe the family history, or attach a scanned pedigree."
          onChange={this.handleChange}
          onValidate={this.validate()}
          formState={this.state.form}
          optional
        />
        
        <Toggle
          field="consanguinity"
          label="Consanguinity"
          onChange={this.handleChange}
          checked={this.state.form.consanguinity === true}
          formState={this.state.form}
        />

        {this.state.form.consanguinity && (
          <TextArea
            field="consanguinityInfo"
            helper="Please provide details"
            onChange={this.handleChange}
            formState={this.state.form}
          />
        )}
            
        {
          this.props.route.isEdited !== true &&
          <FormButton 
          onClick={this.handleBack}
          label="Back"
          back
          >
            Back
          </FormButton> 
        }
      
        <FormButton  
          type="submit" 
          onClick={this.handleConfirm}
        >
          Confirm
        </FormButton> 
      </div>
    );
  }
}

export default ClinicalInfo;