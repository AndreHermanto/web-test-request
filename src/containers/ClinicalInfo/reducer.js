import validator from './../../components/validator';

export function initData(prefilled) {
  var state = {
    form: {
      clinicalInfo: '',
      relevantInvestigation: '',
      familyHistory: '',
      consanguinityInfo: '',
      consanguinity: false
    },
    validationRule: {
      clinicalInfo: 'required'
    },
    validated: false,
    formId: 'clinicalInfoModule'
  };
  
  if(prefilled && Object.keys(prefilled).length !== 0) state.form = prefilled;
  
  // This validates the data in the initial state.
  state.validation = {};
  for (var field in state.validationRule) {
    if(field) {
      state.validation[field] = validator(state.form[field], state.validationRule[field]);
    }
  }
  
  return Object.assign({}, state);
}


/**
 * This sets the form data upon onChange.
 * @param {Object} state Targeted state to be changed.
 * @param {Object} target Target object captured from UI event change.
 */
export function setFormData(state, target) {
  var value, formStateChild;
  switch(target.type) {
    case 'checkbox':
      value = target.checked;
      break;
    default:
      value = target.value;
      break;
  } 

  if(value === false) {
    formStateChild = Object.assign({}, state.form, {
      [target.name]: value,
      consanguinityInfo: ''
    });
  }
  else {
    formStateChild = Object.assign({}, state.form, {
      [target.name]: value
    });
  }
  
  var validationStateChild = Object.assign({}, state.validation, {
    [target.name]: validator(value, state.validationRule[target.name])
  });

  return Object.assign({}, state, {
    form: formStateChild,
    validation: validationStateChild
  });
}

/**
 * Set "validated" state to true - identifying the confirm button is clicked and validation processed.
 * @param {Object} state Targeted state to be changed.
 */
export function validatedToTrue(state) {
  return Object.assign({}, state, {
    validated: true
  });
}