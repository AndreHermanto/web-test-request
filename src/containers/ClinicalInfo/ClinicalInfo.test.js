import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import TestUtils from 'react-addons-test-utils';
import FetchMock from 'fetch-mock';
import ClinicalInfo from './index';

describe('ClinicalInfo: index', () => {
  var props = {
    route: {
      onChange: jest.fn(),
      data: {
        attachments: [
          { id: 1, filename: 'A' },
          { id: 2, filename: 'B' },
          { id: 3, filename: 'C' }
        ],
        consangunity: false
      }
    },
    router:['/step3']
  };
  
  test('renders without crashing', () => {
    const page = renderer.create(React.createElement(ClinicalInfo, props)).toJSON();
    expect(page).toMatchSnapshot();
  });
  
  test('handleChange works - general', () => {
    const page = TestUtils.renderIntoDocument(React.createElement(ClinicalInfo, props));
    const selection = { "target": { 
      "value": "abc",
      "name": "clinicalInfo",
      "type": "text"
    }};
    page.handleChange(selection);
    expect(page.state.form.clinicalInfo).toEqual('abc');  
  });
  
  test('handleChange works - checkbox', () => {
    const page = TestUtils.renderIntoDocument(React.createElement(ClinicalInfo, props));
    const selection = { "target": { 
      "checked": true,
      "name": "consangunity",
      "type": "checkbox"
    }};
    page.handleChange(selection);
    expect(page.state.form.consangunity).toEqual(true);  
  });
  
  test('handleDrop works and adds attachment', async () => {
    FetchMock.post('*', { data: { id: 1 } });
    var page = TestUtils.renderIntoDocument(React.createElement(ClinicalInfo, props));
    var file = [{ name: 'A', preview: ''}];
    await page.handleDrop(file);
    expect(page.state.form.attachments[3].id).toEqual(1);
    FetchMock.restore();
  });
  
  test('handleRemoveAttachment removes an attachment', () => {
    var page = TestUtils.renderIntoDocument(React.createElement(ClinicalInfo, props));
    page.handleRemoveAttachment(0);
    expect(page.state.form.attachments[0].id).toEqual(2);
  });
  
  test('handleBack works', () =>  {
    var page = TestUtils.renderIntoDocument(React.createElement(ClinicalInfo, props));                                         
    page.handleBack();
    expect(page.props.router.pop()).toEqual('/step2')
  });
  
  test('handleNext works', () => {
    var page = TestUtils.renderIntoDocument(React.createElement(ClinicalInfo, props));
    page.handleNext(false);
    expect(page.props.router.pop()).toEqual('/step3')
    page.handleNext(true);
    expect(page.props.router.pop()).toEqual('/step4')
  });
  
  test('handleConfirm works', () => {
    var page = TestUtils.renderIntoDocument(React.createElement(ClinicalInfo, props));
    page.handleConfirm();
    expect(page.state.validated).toEqual(true);
  });
  
  test('closeDeleteModal hides the modal by setting the deleteModal - display state', () => {
    var view = TestUtils.renderIntoDocument(React.createElement(ClinicalInfo, props));                                         
    view.closeDeleteModal();
    expect(view.state.deleteModal.display).toEqual(false);
  });
  
  test('openDeleteModal hides the modal by setting the deleteModal - display state', () => {
    var view = TestUtils.renderIntoDocument(React.createElement(ClinicalInfo, props));
    var select = { currentTarget: { name: 0 } }
    view.openDeleteModal(select);
    expect(view.state.deleteModal.display).toEqual(true);
  });
});
