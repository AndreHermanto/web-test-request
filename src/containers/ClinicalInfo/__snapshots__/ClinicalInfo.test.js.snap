exports[`ClinicalInfo: index renders without crashing 1`] = `
<div>
  <h3
    className="hWxLNB">
    Step 3: Clinical Information
  </h3>
  <div
    className="form-group has-feedback">
    <label
      className="control-label"
      htmlFor={undefined}>
      Provide Clinical Information
      <span
        className="lnMAUb">
        Required
      </span>
    </label>
    <div
      className="kWhIoX">
      Type or copy and paste clinical notes and the diagnosis here. Comprehensive notes increase the chance of a successful diagnosis.
    </div>
    <textarea
      className="ebbddk form-control"
      id={undefined}
      name="clinicalInfo"
      onChange={[Function]}
      style={
        Object {
          "minHeight": 90,
        }
      }
      type={undefined}
      value={undefined} />
    <span
      className="SfhHK help-block" />
  </div>
  <div
    className="form-group has-feedback">
    <label
      className="control-label"
      htmlFor={undefined}>
      Provide results from relevant investigations
      <span
        className="lnMAUb">
        Optional
      </span>
    </label>
    <div
      className="kWhIoX">
      Provide notes from genetic tests, imaging results, etc.
    </div>
    <textarea
      className="ebbddk form-control"
      id={undefined}
      name="relevantInvestigation"
      onChange={[Function]}
      style={
        Object {
          "minHeight": 90,
        }
      }
      type={undefined}
      value={undefined} />
    <span
      className="SfhHK help-block" />
  </div>
  <div
    className="form-group has-feedback">
    <label
      className="control-label"
      htmlFor={undefined}>
      Family History
      <span
        className="lnMAUb">
        Optional
      </span>
    </label>
    <div
      className="kWhIoX">
      Please describe the family history.
    </div>
    <textarea
      className="ebbddk form-control"
      id={undefined}
      name="familyHistory"
      onChange={[Function]}
      style={
        Object {
          "minHeight": 90,
        }
      }
      type={undefined}
      value={undefined} />
    <span
      className="SfhHK help-block" />
  </div>
  <label
    className="jEVCcT">
    Attachments
    <span
      className="lnMAUb">
      Optional
    </span>
  </label>
  <p
    className="kvAVcE"
    style={
      Object {
        "margin": "0 0 4px 0",
      }
    }>
    Please upload any relevant attachments e.g. a scanned pedigree
  </p>
  <div
    className=""
    onClick={[Function]}
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    style={
      Object {
        "border": "1px dashed #aaa",
        "marginBottom": 24,
        "padding": 16,
        "width": "100%",
      }
    }>
    <div
      style={
        Object {
          "marginBottom": 24,
        }
      }>
      <div
        className="iubaNK">
        <button
          className="jEoofW btn btn-link"
          disabled={false}
          onClick={[Function]}
          type="button">
          <span
            className="glyphicon glyphicon-file" />
           
          A
        </button>
        <button
          className="gVOLHi btn btn-xs btn-link"
          disabled={false}
          name={0}
          onClick={[Function]}
          type="button">
          <span
            className="glyphicon glyphicon-trash" />
           Remove
        </button>
      </div>
      <div
        className="iubaNK">
        <button
          className="jEoofW btn btn-link"
          disabled={false}
          onClick={[Function]}
          type="button">
          <span
            className="glyphicon glyphicon-file" />
           
          B
        </button>
        <button
          className="gVOLHi btn btn-xs btn-link"
          disabled={false}
          name={1}
          onClick={[Function]}
          type="button">
          <span
            className="glyphicon glyphicon-trash" />
           Remove
        </button>
      </div>
      <div
        className="iubaNK">
        <button
          className="jEoofW btn btn-link"
          disabled={false}
          onClick={[Function]}
          type="button">
          <span
            className="glyphicon glyphicon-file" />
           
          C
        </button>
        <button
          className="gVOLHi btn btn-xs btn-link"
          disabled={false}
          name={2}
          onClick={[Function]}
          type="button">
          <span
            className="glyphicon glyphicon-trash" />
           Remove
        </button>
      </div>
    </div>
    <button
      className="bBdbjl"
      onClick={[Function]}
      style={
        Object {
          "margin": "12px 0",
        }
      }>
      Upload Attachment
    </button>
    <input
      accept={undefined}
      multiple={true}
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file" />
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Consanguinity
    </label>
    <br />
    <table>
      <tbody>
        <tr>
          <td>
            <div
              className="react-toggle"
              onClick={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}>
              <div
                className="react-toggle-track">
                <div
                  className="react-toggle-track-check">
                  <svg
                    height="11"
                    viewBox="0 0 14 11"
                    width="14">
                    <title>
                      switch-check
                    </title>
                    <path
                      d="M11.264 0L5.26 6.004 2.103 2.847 0 4.95l5.26 5.26 8.108-8.107L11.264 0"
                      fill="#fff"
                      fillRule="evenodd" />
                  </svg>
                </div>
                <div
                  className="react-toggle-track-x">
                  <svg
                    height="10"
                    viewBox="0 0 10 10"
                    width="10">
                    <title>
                      switch-x
                    </title>
                    <path
                      d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
                      fill="#fff"
                      fillRule="evenodd" />
                  </svg>
                </div>
              </div>
              <div
                className="react-toggle-thumb" />
              <input
                checked={false}
                className="react-toggle-screenreader-only"
                name="consanguinity"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                type="checkbox" />
            </div>
          </td>
          <td />
        </tr>
      </tbody>
    </table>
    <span
      className="SfhHK help-block" />
  </div>
  <button
    className="ieDBA"
    disabled={false}
    label="Back"
    onClick={[Function]}>
    Back
  </button>
  <button
    className="bBdbjl"
    disabled={false}
    onClick={[Function]}
    type="submit">
    Next
  </button>
</div>
`;
